File: FVResourceFile.m
//
//  FVResourceFile.m
//  ForkView
//
//  Created by Kevin Wojniak on 5/25/11.
//  Copyright 2011 Kevin Wojniak. All rights reserved.
//

#import "FVResourceFile.h"
#import "FVResourceFilePriv.h"
#import "FVFork.h"
#import "ForkView-Swift.h"

@interface NSError (FVResource)

+ (NSError *)errorWithDescription:(NSString *)description;

@end

@implementation NSError (FVResource)

+ (NSError *)errorWithDescription:(NSString *)description
{
	return [NSError errorWithDomain:@"FVResourceErrorDomain" code:-1 userInfo:[NSDictionary dictionaryWithObject:description forKey:NSLocalizedFailureReasonErrorKey]];
}

@end

struct FVResourceHeader {
	uint32_t dataOffset;
	uint32_t mapOffset;
	uint32_t dataLength;
	uint32_t mapLength;
};

struct FVResourceMap {
	FVResourceHeader headerCopy;
	uint32_t nextMap;
	uint16_t fileRef;
	uint16_t attributes;
	uint16_t typesOffset;
	uint16_t namesOffset;
};

@interface FVResourceFile ()
- (BOOL)readHeader:(FVResourceHeader *)aHeader;
- (BOOL)readMap;
- (BOOL)readTypes;
@end

@implementation FVResourceFile

@synthesize types;

#pragma mark -
#pragma mark Public Methods

- (instancetype)initWithContentsOfURL:(NSURL *)fileURL error:(NSError **)error fork:(FVForkType)forkType
{
	self = [super init];
	if (self == nil) {
		return nil;
	}
	
	fork = [[FVFork alloc] initWithURL:fileURL type:forkType];
	if (!fork) {
		if (error) {
			*error = [NSError errorWithDescription:@"Bad file."];
		}
		return nil;
	}
	
	header = calloc(1, sizeof(FVResourceHeader));
	if ([self readHeader:header] == NO) {
		if (error) {
			*error = [NSError errorWithDescription:@"Invalid header."];
		}
		return nil;
	}
	
	//NSLog(@"HEADER (%u, %u), (%u, %u)", header->dataOffset, header->dataLength, header->mapOffset, header->mapLength);
	
	if ([self readMap] == NO) {
		if (error) {
			*error = [NSError errorWithDescription:@"Invalid map."];
		}
		return nil;
	}
	
	if ([self readTypes] == NO) {
		if (error) {
			*error = [NSError errorWithDescription:@"Invalid types list."];
		}
		return nil;
	}
	
	// Don't open empty (but valid) resource forks
	if (!types || ![types count]) {
		if (error) {
			*error = [NSError errorWithDescription:@"No resources."];
		}
		return nil;
	}
	
	return self;
}

- (instancetype)initWithContentsOfURL:(NSURL *)fileURL error:(NSError **)error
{
	NSError *tmpError = nil;
	FVResourceFile *file = [[[self class] alloc] initWithContentsOfURL:fileURL error:&tmpError fork:FVForkTypeResource];
	if (!file) {
		file = [[[self class] alloc] initWithContentsOfURL:fileURL error:&tmpError fork:FVForkTypeData];
	}
	if (file) {
		return file;
	}
	if (error) {
		*error = tmpError;
	}
	return nil;
}

+ (id)resourceFileWithContentsOfURL:(NSURL *)fileURL error:(NSError **)error
{
	return [(FVResourceFile*)[self alloc] initWithContentsOfURL:fileURL error:error];
}

#pragma mark -
#pragma mark Private Methods

- (void)dealloc
{
	if (header) {
		free(header);
	}
	if (map) {
		free(map);
	}
}

- (BOOL)readHeader:(FVResourceHeader *)aHeader
{
	// read the header values
	if (![fork read:sizeof(aHeader->dataOffset) into:&aHeader->dataOffset] ||
		![fork read:sizeof(aHeader->mapOffset) into:&aHeader->mapOffset] ||
		![fork read:sizeof(aHeader->dataLength) into:&aHeader->dataLength] ||
		![fork read:sizeof(aHeader->mapLength) into:&aHeader->mapLength]) {
		return NO;
	}
	
	// swap from big endian to host
	aHeader->dataOffset = CFSwapInt32BigToHost(aHeader->dataOffset);
	aHeader->mapOffset = CFSwapInt32BigToHost(aHeader->mapOffset);
	aHeader->dataLength = CFSwapInt32BigToHost(aHeader->dataLength);
	aHeader->mapLength = CFSwapInt32BigToHost(aHeader->mapLength);
	
	unsigned fileLength = [fork length];
	if ((aHeader->dataOffset + aHeader->dataLength > fileLength) || (aHeader->mapOffset + aHeader->mapLength > fileLength)) {
		return NO;
	}
	
	return YES;
}

- (BOOL)readMap
{
	// seek to the map offset
	if (![fork seekTo:header->mapOffset]) {
		return NO;
	}
	
	// read the map values
	map = calloc(1, sizeof(FVResourceMap));
	if ([self readHeader:&map->headerCopy] == NO) {
		return NO;
	}
	
	// the map contains a copy of the header, or all 0s
	char zeros[16] = {0};
	if (memcmp(&map->headerCopy, header, sizeof(FVResourceHeader)) != 0 && memcmp(&map->headerCopy, zeros, sizeof(zeros)) != 0) {
		printf("Bad match!\n");
	}
	
	if (![fork read:sizeof(map->nextMap) into:&map->nextMap] ||
		![fork read:sizeof(map->fileRef) into:&map->fileRef] ||
		![fork read:sizeof(map->attributes) into:&map->attributes] ||
		![fork read:sizeof(map->typesOffset) into:&map->typesOffset] ||
		![fork read:sizeof(map->namesOffset) into:&map->namesOffset]) {
		return NO;
	}
	
	// swap from big endian to host
	map->nextMap = CFSwapInt32BigToHost(map->nextMap);
	map->fileRef = CFSwapInt16BigToHost(map->fileRef);
	map->attributes = CFSwapInt16BigToHost(map->attributes);
	map->typesOffset = CFSwapInt16BigToHost(map->typesOffset);
	map->namesOffset = CFSwapInt16BigToHost(map->namesOffset);
	
	//NSLog(@"MAP (%u, %u, %x, %u, %u)", map->nextMap, map->fileRef, map->attributes, map->typesOffset, map->namesOffset);
	
	return YES;
}

- (BOOL)readTypes
{
	unsigned typesOffset = [fork position];
	
	uint16_t numberOfTypes;
	if (![fork read:sizeof(numberOfTypes) into:&numberOfTypes]) {
		return NO;
	}
	numberOfTypes = CFSwapInt16BigToHost(numberOfTypes) + 1;
	
	NSMutableArray *typesTemp = [NSMutableArray array];
	
	for (uint16_t typeIndex=0; typeIndex<numberOfTypes; typeIndex++) {
		uint32_t type;
		uint16_t numberOfResources;
		uint16_t referenceListOffset;
		
		if (![fork read:sizeof(type) into:&type] ||
			![fork read:sizeof(numberOfResources) into:&numberOfResources] ||
			![fork read:sizeof(referenceListOffset) into:&referenceListOffset]) {
			return NO;
		}
		
		type = CFSwapInt32BigToHost(type);
		numberOfResources = CFSwapInt16BigToHost(numberOfResources) + 1;
		referenceListOffset = CFSwapInt16BigToHost(referenceListOffset);
		
		FVResourceType *obj = [[FVResourceType alloc] init];
		obj.type = type;
		obj.count = numberOfResources;
		obj.offset = referenceListOffset;
		[typesTemp addObject:obj];
		
		//NSLog(@"[%u]: %@, %u, %u", typeIndex+1, NSFileTypeForHFSTypeCode(type), numberOfResources, referenceListOffset);
	}
	
	for (FVResourceType *obj in typesTemp) {
		NSMutableArray *resourcesTemp = [NSMutableArray array];
		
		for (uint16_t resIndex=0; resIndex<obj.count; resIndex++) {
			if (![fork seekTo:typesOffset + obj.offset + (resIndex * 12)]) {
				return NO;
			}
			
			uint16_t resourceID;
			int16_t nameOffset;
			uint8_t attributes;
			uint8_t dataOffsetBytes[3];
			uint32_t resHandle;
			if (![fork read:sizeof(resourceID) into:&resourceID] ||
				![fork read:sizeof(nameOffset) into:&nameOffset] ||
				![fork read:sizeof(attributes) into:&attributes] ||
				![fork read:sizeof(dataOffsetBytes) into:&dataOffsetBytes] ||
				![fork read:sizeof(resHandle) into:&resHandle]) {
					return NO;
			}
			
			resourceID = CFSwapInt16BigToHost(resourceID);
			nameOffset = CFSwapInt16BigToHost(nameOffset);
			resHandle = CFSwapInt32BigToHost(resHandle);
			
			uint32_t dataOffset = 0;
			dataOffset |= dataOffsetBytes[0] << 16;
			dataOffset |= dataOffsetBytes[1] << 8;
			dataOffset |= dataOffsetBytes[2];
			
			char name[256];
			uint8_t nameLength = 0;
			if ((nameOffset != -1) && ([fork seekTo:header->mapOffset + map->namesOffset + nameOffset])) {
				if (![fork read:sizeof(nameLength) into:&nameLength] || ![fork read:nameLength into:name]) {
					nameLength = 0;
				}
			}
			name[nameLength] = 0;
			
			uint32_t dataLength = 0;
			if ([fork seekTo:header->dataOffset + dataOffset] && [fork read:sizeof(dataLength) into:&dataLength]) {
				dataLength = CFSwapInt32BigToHost(dataLength);
			}
			
			//NSLog(@"%@[%u] %u %s", obj.typeString, resourceID, dataLength, name);
			FVResource *resource = [[FVResource alloc] init];
            resource.ident = resourceID;
            resource.dataSize = dataLength;
            resource.dataOffset = dataOffset + sizeof(dataOffset);
			if (strlen(name)) {
				[resource setName:[NSString stringWithCString:name encoding:NSMacOSRomanStringEncoding]];
			}
			resource.file = self;
			resource.type = obj;
			[resourcesTemp addObject:resource];
		}
		
		NSSortDescriptor *sort = [[NSSortDescriptor alloc] initWithKey:@"ident" ascending:YES];
		[resourcesTemp sortUsingDescriptors:[NSArray arrayWithObject:sort]];
		
		obj.resources = resourcesTemp;
	}		
	
	NSSortDescriptor *sort = [[NSSortDescriptor alloc] initWithKey:@"typeString" ascending:YES selector:@selector(caseInsensitiveCompare:)];
	[typesTemp sortUsingDescriptors:[NSArray arrayWithObject:sort]];
	
	types = typesTemp;
	
	return YES;
}

- (NSData *)dataForResource:(FVResource *)resource
{
	if (![fork seekTo:header->dataOffset + resource.dataOffset]) {
		return nil;
	}
	NSMutableData *data = [NSMutableData dataWithLength:resource.dataSize];
	if (!data) {
		return nil;
	}
	if (![fork read:resource.dataSize into:[data mutableBytes]]) {
		return nil;
	}
	return data;
}

- (FVForkType)forkType
{
	return fork.type;
}

@end

File: FVResourceFile.swift
//
//  FVResourceFile.swift
//  ForkView
//
//  Created by C.W. Betts on 5/12/15.
//

import Foundation

extension NSError {
    fileprivate class func errorWithDescription(_ description: String) -> NSError {
        return NSError(domain: "FVResourceErrorDomain", code: -1, userInfo: [NSLocalizedFailureReasonErrorKey: description])
    }
}

private func ==(lhs: FVResourceFile.ResourceHeader, rhs: FVResourceFile.ResourceHeader) -> Bool {
    if lhs.dataOffset != rhs.dataOffset {
        return false
    } else if lhs.mapOffset != rhs.mapOffset {
        return false
    } else if lhs.dataLength != rhs.dataLength {
        return false
    } else if lhs.mapLength != rhs.mapLength {
        return false
    }

    return true
}

final public class FVResourceFile: NSObject {
    @objc var types: [FVResourceType] = []
    @objc var isResourceFork = false

    private var header = ResourceHeader()
    private var map = ResourceMap()
    private let dataReader: FVDataReader!

    fileprivate struct ResourceHeader: Equatable {
        var dataOffset: UInt32 = 0
        var mapOffset: UInt32 = 0
        var dataLength: UInt32 = 0
        var mapLength: UInt32 = 0
    }

    private struct ResourceMap {
        var headerCopy = ResourceHeader()
        var nextMap: UInt32 = 0
        var fileRef: UInt16 = 0
        var attributes: UInt16 = 0
        var typesOffset: UInt16 = 0
        var namesOffset: UInt16 = 0
    }

    private func readHeader(_ aHeader: inout ResourceHeader) -> Bool {
        // read the header values
        if !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: aHeader.dataOffset)), into: &aHeader.dataOffset) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: aHeader.mapOffset)), into: &aHeader.mapOffset) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: aHeader.dataLength)), into: &aHeader.dataLength) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: aHeader.mapLength)), into: &aHeader.mapLength) {
                return false
        }

        // swap from big endian to host
        aHeader.dataOffset = aHeader.dataOffset.bigEndian
        aHeader.mapOffset = aHeader.mapOffset.bigEndian
        aHeader.dataLength = aHeader.dataLength.bigEndian
        aHeader.mapLength = aHeader.mapLength.bigEndian

        let fileLength = UInt32(dataReader.length)
        if (aHeader.dataOffset + aHeader.dataLength > fileLength) || (aHeader.mapOffset + aHeader.mapLength > fileLength) {
            return false
        }

        return true
    }

    private init(contentsOfURL fileURL: NSURL, resourceFork: Bool) throws {
        if let dataReader = FVDataReader(url: fileURL, resourceFork: resourceFork) {
            self.dataReader = dataReader
            super.init()

            if !readHeader(&header) {
                throw NSError.errorWithDescription("Invalid header.")
            }

            //NSLog(@"HEADER (%u, %u), (%u, %u)", header->dataOffset, header->dataLength, header->mapOffset, header->mapLength);

            if !readMap() {
                throw NSError.errorWithDescription("Invalid map.")
            }

            if !readTypes() {
                throw NSError.errorWithDescription("Invalid types list.")
            }

            // Don't open empty (but valid) resource forks
            if types.count == 0 {
                throw NSError.errorWithDescription("No resources.")
            }
            return
        } else {
            self.dataReader = nil
            super.init()

            throw NSError.errorWithDescription("Bad File.")
        }
    }

    private func readMap() -> Bool {
        // seek to the map offset
        if !dataReader.seekTo(Int(header.mapOffset)) {
            return false
        }

        // read the map values
        if !readHeader(&map.headerCopy) {
            return false
        }

        let zeros = [Int8](repeating: 0, count: 16)
        if (map.headerCopy != header) && (memcmp(&map.headerCopy, zeros, zeros.count) != 0) {
            print("Bad match!")
        }

        if !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: map.nextMap)), into: &map.nextMap) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: map.fileRef)), into: &map.fileRef) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: map.attributes)), into: &map.attributes) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: map.typesOffset)), into: &map.typesOffset) ||
            !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: map.namesOffset)), into: &map.namesOffset) {
                return false
        }

        map.nextMap = map.nextMap.bigEndian
        map.fileRef = map.fileRef.bigEndian
        map.attributes = map.attributes.bigEndian
        map.typesOffset = map.typesOffset.bigEndian
        map.namesOffset = map.namesOffset.bigEndian

        return true
    }

    private func readTypes() -> Bool {
        let typesOffset = dataReader.position

        var numberOfTypes: UInt16 = 0
        if !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: numberOfTypes)), into: &numberOfTypes) {
            return false
        }
        numberOfTypes = numberOfTypes.bigEndian + 1

        var typesTemp = [FVResourceType]()
        for _ in 0..<numberOfTypes {
            var type: OSType = 0
            var numberOfResources: UInt16 = 0
            var referenceListOffset: UInt16 = 0
            if !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: type)), into: &type) ||
                !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: numberOfResources)), into: &numberOfResources) ||
                !dataReader.read(CUnsignedInt(MemoryLayout.size(ofValue: referenceListOffset)), into: &referenceListOffset) {
                    return false
            }
            type = type.bigEndian
            numberOfResources = numberOfResources.bigEndian + 1
            referenceListOffset = referenceListOffset.bigEndian

            let obj = FVResourceType()
            obj.type = type
            obj.count = UInt32(numberOfResources)
            obj.offset = UInt32(referenceListOffset)
            typesTemp.append(obj)
        }

        for obj in typesTemp {
            var tmpResources = [FVResource]()

            for resIndex in 0..<obj.count {
                if !dataReader.seekTo(typesOffset + Int(obj.offset) + Int(resIndex * 12)) {
                    return false
                }

                var resourceID: UInt16 = 0
                var nameOffset: Int16 = 0
                var attributes: UInt8 = 0
                var dataOffsetBytes: (UInt8, UInt8, UInt8) = (0, 0, 0)
                var resHandle: UInt32 = 0

                if !dataReader.read(UInt32(MemoryLayout.size(ofValue: resourceID)), into: &resourceID) ||
                    !dataReader.read(UInt32(MemoryLayout.size(ofValue: nameOffset)), into: &nameOffset) ||
                    !dataReader.read(UInt32(MemoryLayout.size(ofValue: attributes)), into: &attributes) ||
                    !dataReader.read(UInt32(MemoryLayout.size(ofValue: dataOffsetBytes)), into: &dataOffsetBytes) ||
                    !dataReader.read(UInt32(MemoryLayout.size(ofValue: resHandle)), into: &resHandle) {
                        return false
                }

                resourceID = resourceID.bigEndian
                nameOffset = nameOffset.bigEndian
                resHandle = resHandle.bigEndian

                let dataOffset: UInt32 = {
                    var toRet = UInt32(dataOffsetBytes.2)
                    toRet |= UInt32(dataOffsetBytes.0) << 16
                    toRet |= UInt32(dataOffsetBytes.1) << 8

                    return toRet
                }()

                var name = [Int8](repeating: 0, count: 256)
                var nameLength: UInt8 = 0

                if (nameOffset != -1) && (dataReader.seekTo(Int(header.mapOffset) + Int(map.namesOffset) + Int(nameOffset))) {
                    if !dataReader.read(UInt32(MemoryLayout.size(ofValue: nameLength)), into: &nameLength) || !dataReader.read(UInt32(nameLength), into: &name) {
                        nameLength = 0
                    }
                }
                name[Int(nameLength)] = 0

                var dataLength: UInt32 = 0
                if dataReader.seekTo(Int(header.dataOffset + dataOffset)) && dataReader.read(UInt32(MemoryLayout.size(ofValue: dataLength)), into: &dataLength) {
                    dataLength = dataLength.bigEndian
                }

                //NSLog(@"%@[%u] %u %s", obj.typeString, resourceID, dataLength, name);
                let resource = FVResource()
                resource.ident = resourceID
                resource.dataSize = dataLength
                resource.dataOffset = dataOffset + UInt32(MemoryLayout.size(ofValue: dataOffset))
                if strlen(name) != 0 {
                    resource.name = String(cString: name, encoding: String.Encoding.macOSRoman)!
                }
                resource.file = self
                resource.type = obj
                tmpResources.append(resource)
            }
            tmpResources.sort { (lhs, rhs) -> Bool in
                return lhs.ident > rhs.ident
            }
            obj.resources = tmpResources as NSArray
        }

        typesTemp.sort { (lhs, rhs) -> Bool in
            let compVal = lhs.typeString.caseInsensitiveCompare(rhs.typeString)
            return compVal == .orderedAscending
        }

        types = typesTemp
        return true
    }

    internal func dataForResource(resource: FVResource) -> NSData? {
        if !dataReader.seekTo(Int(header.dataOffset + resource.dataOffset)) {
            return nil
        }
        if let data = NSMutableData(length: Int(resource.dataSize)) {
            if !dataReader.read(resource.dataSize, into: data.mutableBytes) {
                return nil
            }
            return data
        }
        return nil
    }

    // TODO: implement, but how?
    //convenience init?(contentsOfURL fileURL: NSURL, error: NSErrorPointer) {
    //
    //}

    public class func resourceFileWithContentsOfURL(fileURL: URL) throws -> FVResourceFile {
        let url = fileURL as NSURL
        do {
            let file = try FVResourceFile(contentsOfURL: url, resourceFork: true)
            file.isResourceFork = true
            return file
        } catch let error1 {
            var tmpError: Error? = error1
            do {
                let file = try FVResourceFile(contentsOfURL: url, resourceFork: false)
                return file
            } catch let error {
                tmpError = error
            }
            throw tmpError!
        }
    }
}
